File: advent_of_code/python/2022/day06.py
test_cases = [
    'mjqjpqmgbljsphdztnvjfqwrcgsmlb',
    'bvwbjplbgvbhsrlpgdmjqwftvncz',
    'nppdvjthqldpwncqszvftbrmjlhg',
    'nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg',
    'zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw'
]


def readfile(filename):
    s = None
    with open(filename) as f:
        s = f.readline().strip()
    return s


def day06_part1(s):
    n = 4
    for i in range(len(s)-n):
        cs = set(s[i:i+n])
        if len(cs) == n:
            return i + n


def day06_part2(s):
    n = 14
    for i in range(len(s)-n):
        cs = set(s[i:i+n])
        if len(cs) == n:
            return i + n


if __name__ == '__main__':
    input_filename = 'input.txt'
    testing = False

    if testing:
        print('Part 1:')
        for i, test in enumerate(test_cases):
            r = day06_part1(test)
            print('- Test Case %s: %s' % (i, r))
        print()
        print('Part 2:')
        for i, test in enumerate(test_cases):
            r = day06_part2(test)
            print('- Test Case %s: %s' % (i, r))
    else:
        s = readfile(input_filename)
        result1 = day06_part1(s)
        print('Result1: %s' % result1)
        result2 = day06_part2(s)
        print('Result2: %s' % result2)

File: go/aoc/2022/day06.go
package main

import (
	"fmt"
	"os"
)

func part1(input string) int {
outer:
	for i := 4; i < len(input); i++ {
		for j := i - 4; j < i-1; j++ {
			for k := j + 1; k < i; k++ {
				if input[j] == input[k] {
					continue outer
				}
			}
		}
		return i
	}
	return -1
}

func part2(input string) int {
	const n = 14
outer:
	for i := n; i < len(input); i++ {
		for j := i - n; j < i-1; j++ {
			for k := j + 1; k < i; k++ {
				if input[j] == input[k] {
					continue outer
				}
			}
		}
		return i
	}
	return -1
}

var tests []string = []string{
	"mjqjpqmgbljsphdztnvjfqwrcgsmlb",
	"bvwbjplbgvbhsrlpgdmjqwftvncz",
	"nppdvjthqldpwncqszvftbrmjlhg",
	"nznrnfrfntjfmvfwmzdfjlvtqnbhcprsg",
	"zcfzfwzzqfrljwzlrfnpqdbhtmscgvjw",
}

func main() {
	const DEBUG = true
	filename := "input.txt"
	var input string
	if DEBUG {
		fmt.Println("Part 1:")
		for _, test := range tests {
			fmt.Printf("%s = %d\n", test, part1(test))
		}
		fmt.Println("\nPart 2:")
		for _, test := range tests {
			fmt.Printf("%s = %d\n", test, part2(test))
		}
	} else {
		s, err := os.ReadFile(filename)
		if err != nil {
			panic(err)
		}
		input = string(s)

		fmt.Printf("2022, Day 6, part 1: %v\n", part1(input))
		fmt.Printf("2022, Day 6, part 2: %v\n", part2(input))
	}
}
